### ISSUE TITLE
InlinePanel alway creates new objects

### ISSUE THREAD

**Issue by nomeata** [torchbox/wagtail#1853]
Hi,

I have a simple setup like this:
```
class APage(Page):
    content_panels = Page.content_panels + [
        InlinePanel('ordnungen', label="Ordnungen"),
    ]


class Ordnung(Orderable):
    page = ParentalKey(APage, related_name='ordnungen')
    ordnung = models.CharField(max_length=250)
    panels = [
            FieldPanel('ordnung')
    ]
```

As expected, I can create a  `APage`, and edit the set of `Ordnung` there. The problem is that whenever I publish the page, wagtail (or rather modelcluster) deletes all existing objects and creates new ones from the content in the forms. This can be observed by looking at the ids of the objects, and obviously breaks anything that references these objects.

I can work around the issue by issuing `delete * from wagtailcore_pagerevision`. Also, a bit of naive `print()` and `traceback` debugging in modelcluster shows me that wagtail always adds these objects from the JSON representation, and _never_ from the DB. The data in `wagtailcore_pagerevision` always has `pk: null`; I would expect that to be a valid id after the first publish. Indeed, after the work-around (`delete * from wagtailcore_pagerevision`), the next time I edit the page, the data _is_ read from db, the objects are now stored _with_ `pk` in the JSON, and further edits to _these_ object preserve the data base objects. Newly added objects, though, still show the old recreate-on-every-publish behaviour.

My bet is that wagtail, when editing a page, re-uses the last entry in `wagtailcore_pagerevision` to populate the forms and managers and what not, and my conclusion is that it should do that only for pages that are in “draft” mode, and not for those that have been published previous. This way, in the latter case, the current `pk` values find their way into the JSON.

**Comment by nomeata** [comment]
Or maybe the problem is that the revision that is being saved when the page is published for some reason does not contain the primary keys of the children objects. If they were there, then loading the latest revision (instead of the page directly) would work as well. Although this redudancy of data (db content vs. latest page revision) makes me uneasy.

**Comment by nomeata** [comment]
Indeed, `Page.save_revision()` is run _before_ the children objects are written to the data base – no surprise the revision does not know about the pk yet. But the code to publish a page needs to recreate the revision first, in order to run `.publish()` on it, which in turn runs `page.save()`. So maybe, after doing that, the revision needs to be updated with any primary keys?

Anyways, I have dug deep enough for a wagtail user (or rather, someone trying to help a desparate wagtail user), and I hope with this analysis you know what to do to fix this. Thanks!

**Comment by gasman** [comment]
@nomeata Thanks for the report and the detective work! [django-modelcluster](https://github.com/torchbox/django-modelcluster) (the underlying component that handles previews/draft revisions) is designed to preserve existing objects just like Django's own modelformsets do, so it's unfortunate that Wagtail is failing to do so. I suspect this bug was introduced in #566 if not earlier, so it's been around for a while...

In the simple case (publishing directly from the edit form) it would indeed make sense to publish first - creating the real database objects - and create the revision from that. That won't work for pages that go through the moderation workflow, though:

* a page (with newly-added child objects without IDs) is submitted for moderation
* it then receives further draft edits (possibly including edits to those child objects)
* the original revision is then approved by the moderator, and published; the child objects receive IDs at this point
* we then have to somehow attach those IDs to the later revisions

I guess the only way of doing this reliably is to generate our own object IDs when the objects are first created. And that would be a major change to the logic :-/

**Comment by nomeata** [comment]
How about this to mitigate the problem: When creating a new revision, and the revision to base the new one on (usually the latest?) is the currently published revision, then do not use the serialized data therein, but read it from the real objects.

This way, the bug will only be observed if, in your sequence of action, a _new_ inline object is created, and something else on the site references it, before the later.revision is published.

In particular, with my proposed fix, one a object is created and there are no more unpublished revisions based on tlan earlier unpublished draft, then this object stays around without problems, as all revisions that are created later, independent of their publishing order, know the correct id. This would work good enough for something long-lived like categories or tags.

**Comment by gasman** [comment]
@nomeata Good idea! We've actually considered making that change for a separate reason (but haven't got round to it yet): when people make edits to pages directly at the model level (as an automated data import, for example) it's not obvious that they also need to update the revisions, causing confusion when the edit interface appears to show an old version. I'll go ahead and make that change now, since there are several good reasons to do so.

**Comment by nomeata** [comment]
Great, thanks :-)

**Comment by gasman** [comment]
Have now implemented your proposed fix in #1881 - will keep this ticket open as we still need to account for the moderation workflow scenario, but that's a lower priority fix.
